core/src/main/java/org/apache/iceberg/RemoveSnapshots.java

```diff
@@ -208,12 +226,40 @@ private void cleanExpiredFiles(List<Snapshot> snapshots, Set<Long> validIds, Set
       }
     }
 
+    // find manifests to clean up that were only referenced by snapshots that have expired
     Set<String> manifestListsToDelete = Sets.newHashSet();
     Set<String> manifestsToDelete = Sets.newHashSet();
     Set<ManifestFile> manifestsToRevert = Sets.newHashSet();
     for (Snapshot snapshot : base.snapshots()) {
       long snapshotId = snapshot.snapshotId();
       if (!validIds.contains(snapshotId)) {
+        // determine whether the changes in this snapshot are in the current table state
+        if (pickedAncestorSnapshotIds.contains(snapshotId)) {
+          // this snapshot was cherry-picked into the current table state, so skip cleaning it up. its changes will
+          // expire when the picked snapshot expires.
+          // A -- C (source=B)
+          //  `- B <-- this commit
+          continue;
+        }
+
+        long sourceSnapshotId = PropertyUtil.propertyAsLong(
+            snapshot.summary(), SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP, -1);
+        if (ancestorIds.contains(sourceSnapshotId)) {
+          // this commit was cherry-picked from a commit that is in the current table state. do not clean up its
+          // changes because it would revert data file additions that are in the current table.
+          // A -- B
+          //  `- C (source=B) <-- this commit
```

Comment:
Quick note: this case needs a follow-up because we could expire A and B, then later expire C (not in the current table state) and not be able to detect that we should not delete its files. Expiring B means that we need to also expire C.

Reply:
NOTE: I think we can handle this case, i.e. when picking a snapshot for expiration, we should also pick its source snapshot, but considering the gotchas, I would recommend to keep the API simple with expireOlderThan. Also our use case is to use `expireOlderThan` API for now.

I will log a ticket for this.